pyro/poutine/enumerate_messenger.py

```diff
@@ -5,13 +5,68 @@
 from .runtime import _ENUM_ALLOCATOR
 
 
+def _tmc_sample(msg):
+    dist, num_samples = msg["fn"], msg["infer"].get("num_samples")
+
+    # find batch dims that aren't plate dims
+    batch_shape = [1] * len(dist.batch_shape)
+    for f in msg["cond_indep_stack"]:
+        if f.vectorized:
+            batch_shape[f.dim] = f.size
+    batch_shape = tuple(batch_shape)
+
+    # sample a batch
+    sample_shape = (num_samples,) if batch_shape == dist.batch_shape else ()
+    fat_sample = dist(sample_shape=sample_shape)  # TODO thin before sampling
```

Comment:
what'd the TODO about?

Reply:
If a sample site depends on `N > 1` upstream sample sites that have been multiply sampled, `fat_sample` will contain `num_samples ** (N-1)` too many samples, which are then discarded in the `while` loop below this line.  A more efficient implementation would apply the logic in this function to the parameters of `dist` before sampling, but doing this generically would require extra distribution plumbing.